### PR TITLE
[Merged by Bors] - Xtask CI

### PR DESCRIPTION
**Pull request by CleanCut**
This PR is easiest to review commit by commit.

Followup on https://github.com/bevyengine/bevy/pull/1309#issuecomment-767310084

- [x] Switch from a bash script to an xtask rust workspace member.
  - Results in ~30s longer CI due to compilation of the xtask itself
  - Enables Bevy contributors on any platform to run `cargo ci` to run linting -- if the default available Rust is the same version as on CI, then the command should give an identical result.
- [x] Use the xtask from official CI so there's only one place to update.
- [x] Bonus: Run clippy on the _entire_ workspace (existing CI setup was missing the `--workspace` flag
  - [x] Clean up newly-exposed clippy errors 

~#1388 builds on this to clean up newly discovered clippy errors -- I thought it might be nicer as a separate PR.~  Nope, merged it into this one so CI would pass.

### CODE REVIEW

**Comment by CleanCut** [comment]
❓ Odd, this works for me locally with both stable and nightly...but CI doesn't seem to find the .cargo/config file.  I'll look into this tomorrow night unless someone beats me to it.

File: tools/ci/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "ci"
+version = "0.1.0"
+authors = ["Nathan Stocks <nathan@agileperception.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+xshell = "0.1"
```
Comment (bjorn3) [inline tools/ci/Cargo.toml:10]: nit: missing trailing newline

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -35,13 +35,8 @@ jobs:
...
       - name: Check the format
-        run: cargo fmt --all -- --check
-        if: runner.os == 'linux' && matrix.toolchain == 'stable'
-
-      # type complexity must be ignored because we use huge templates for queries
-      # -A clippy::manual-strip: strip_prefix support was added in 1.45. we want to support earlier rust versions
-      - name: Clippy
-        run: cargo clippy --all-targets --all-features -- -D warnings -A clippy::type_complexity -A clippy::manual-strip
+        # See tools/ci/src/main.rs for the commands this runs
+        run: cargo ci
```
Comment (bjorn3) [inline .github/workflows/ci.yml:39]: This makes it a bit slower to find the failed step.

Comment (CleanCut) [reply .github/workflows/ci.yml:39]: In what way?

Comment (CleanCut) [reply .github/workflows/ci.yml]: Ah, I see it. It took 38s to compile the xtask in [the first run](https://github.com/bevyengine/bevy/pull/1387/checks?check_run_id=1828137711#step:5:386), and 24s in [the second run](https://github.com/bevyengine/bevy/pull/1387/checks?check_run_id=1828156485#step:5:194).  So, that makes the `build (stable, ubuntu-latest)` job ~30s longer. Looking back at the 5 most recent passing builds before this, that is also typically the longest job (which makes sense -- it's the only one we do `cargo fmt` and `cargo clippy` in), so I expect this there will generally be a ~30s regression in exchange for the cross-platform portability, integration, and Rusty-ness of this approach. 😉

File: tools/ci/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "ci"
+version = "0.1.0"
+authors = ["Nathan Stocks <nathan@agileperception.com>"]
```
Comment (MinerSebas) [inline tools/ci/Cargo.toml:4]:
```suggestion
authors = [
  "Bevy Contributors <bevyengine@gmail.com>",
  "Carter Anderson <mcanders1@gmail.com>",
]
```
The authors entry should be the same as all other crates in the Repo.

Comment (CleanCut) [reply tools/ci/Cargo.toml]: Good point. I updated the `authors` entry as per [the pattern for wholly-contributed crates](https://github.com/bevyengine/bevy/blob/b71ada03465a0343b4ead1b2896f44587e719a47/crates/bevy_tasks/Cargo.toml#L5-L9) that I see.

**Comment by cart** [comment]
I dig this. I'm definitely looking forward to consistent (and easy to run) ci validation. Resolve merge conflicts and we're good to go!.

**Comment by CleanCut** [comment]
✅ Resolved

**Pull request closed without merging**